--- a/xbmc/music/spotyXBMC/thumb/ThumbStore.cpp.org	2012-09-09 21:47:41.000000000 +0100
+++ B/xbmc/music/spotyXBMC/thumb/ThumbStore.cpp	2012-09-09 21:48:48.000000000 +0100
@@ -28,7 +28,7 @@
 #include <string.h>
 #include "URL.h"
 #include "../../../filesystem/FileCurl.h"
-#include "tinyXML/tinyxml.h"
+#include "../../../pvrclients/MediaPortal/lib/tinyxml/tinyxml.h"
 #include <fstream>
 #include "../../../filesystem/File.h"
 
