diff --git a/xbmc/music/spotyXBMC/SxSettings.cpp b/xbmc/music/spotyXBMC/SxSettings.cpp
index 3827c61..6fe69bc 100644
--- a/xbmc/music/spotyXBMC/SxSettings.cpp
+++ b/xbmc/music/spotyXBMC/SxSettings.cpp
@@ -136,7 +136,7 @@
     		{ addon->GetSetting(radioC + "2") == "true", 0x2 },
     		{ addon->GetSetting(radioC + "3") == "true", 0x4 },
     		{ addon->GetSetting(radioC + "4") == "true", 0x8 },
-    		{ addon->GetSetting(radioC + "") == "true", 0x10 },
+    		{ addon->GetSetting(radioC + "5") == "true", 0x10 },
     		{ addon->GetSetting(radioC + "6") == "true", 0x20 },
     		{ addon->GetSetting(radioC + "7") == "true", 0x40 },
     		{ addon->GetSetting(radioC + "8") == "true", 0x80 },
@@ -150,7 +150,7 @@
     		{ addon->GetSetting(radioC + "16") == "true", 0x8000 },
     		{ addon->GetSetting(radioC + "17") == "true", 0x10000 },
     		{ addon->GetSetting(radioC + "18") == "true", 0x20000 },
-			  { addon->GetSetting(radioC + "19") == "true", 0x40000 },
+     		{ addon->GetSetting(radioC + "19") == "true", 0x40000 },
     		{ addon->GetSetting(radioC + "20") == "true", 0x80000 },
     		{ addon->GetSetting(radioC + "21") == "true", 0x100000 },
     		{ addon->GetSetting(radioC + "22") == "true", 0x200000 },
